### ISSUE TITLE
[new rule] enforce inject() at the top of the class to prevent runtime bugs

### ISSUE THREAD

**Issue by AdiMarianMutu** [angular-eslint/angular-eslint#3008]
### Summary

I'd like to propose a new rule, `inject-at-top`, that enforces every `inject()` call to be declared at the very top of the class, before any other member.

This is not a stylistic rule. Its purpose is to prevent a real runtime bug that is easy to introduce and silently bypasses the TypeScript compiler.

### The bug

Class fields are initialized in the order they are written. That means `this.someService` only exists from the moment the line `private readonly someService = inject(SomeService);` has executed. Anything declared above that line which tries to read from `this.someService` will see `undefined` at construction time.

TypeScript does catch this when an earlier field references a later field *directly* (TS2729: "used before its initialization"). What it does not catch is the same bug behind one level of indirection, when the earlier initializer calls a getter or a method that itself reaches for the injected service. TS2729 is a local, shallow check: it does not walk into a getter body or a method body, so the moment the read of `this.someService` happens behind that indirection the compiler lets it through and the failure only shows up at runtime. Moving the getter or the method above or below the field that uses it does not help, the compiler does not trace through them either way.

Consider this component:

```ts
@Component({ /* ... */ })
class MyComponent {
  protected readonly pageTitle = `Dashboard ${this.fullName}`;

  private get fullName() {
    return `${this.userService.firstName} ${this.userService.lastName}`.trim();
  }

  private readonly userService = inject(UserService);
}
```

This compiles. The IDE shows no errors. TypeScript is happy. But at runtime, the `pageTitle` initializer runs first, calls the `fullName` getter, which dereferences `this.userService`, and `userService` has not been assigned yet. The class blows up with `Cannot read properties of undefined (reading 'firstName')`. The error points at the getter, not at the real cause (the `inject()` call sitting below it), and there is nothing about the code that looks obviously wrong.

Inlining the getter isn't really an option either: `fullName` composes two fields and trims the result, which is the kind of thing you actually want behind a name. The moment any non-trivial formatting, null-handling or permission check lives behind a getter or a helper method, you are one innocent reorder away from the same crash.

### The proposal

A new rule `@angular-eslint/inject-at-top` that:

- Applies to classes decorated with `@Component`, `@Directive`, `@Injectable` or `@Pipe`.
- Walks the class body in source order and reports any property initialized with `inject(...)` that appears after another (non-static) member.
- Static members are ignored because they are initialized at class-definition time, not per instance, so their position relative to `inject()` is irrelevant.
- No autofix: the correct position for each `inject()` call depends on the grouping the author wants at the top of the class, and auto-moving them risks shuffling unrelated members or comments. A clear error message is enough.

With the rule in place, the example above fails lint on the `inject(UserService)` line, with a message telling the author to move it to the top of the class, which by definition removes any chance of another member being initialized before it.

### Why it should be in `recommended`

The bug it catches is a true runtime bug, not a convention. The TypeScript compiler cannot see it once it is behind a getter or a method, and there is no reason in idiomatic Angular code to have an `inject()` call sit anywhere other than the top of the class. `prefer-inject` is already in `recommended` and nudges people toward the field-level `inject()` pattern, this rule closes the ordering footgun that comes with it.

### Scope

- In scope: field-level `inject()` calls inside `@Component` / `@Directive` / `@Injectable` / `@Pipe` classes.
- Out of scope (at least for the initial version): `inject()` calls inside functions or factories, non-Angular classes, and detecting which specific consumer would fail, the rule keeps it simple and just enforces position.

**Comment by reduckted** [comment]
Without creating a whole new rule for this specific case, could you use [`eslint-plugin-perfectionist/sort-classes`](https://perfectionist.dev/rules/sort-classes)?

You wouldn't be able to make it _specifically_ put inject-initialized properties first, but you would be able to configure it to put properties before getters and methods.

**Comment by AdiMarianMutu** [comment]
> Without creating a whole new rule for this specific case, could you use [`eslint-plugin-perfectionist/sort-classes`](https://perfectionist.dev/rules/sort-classes)?
> 
> You wouldn't be able to make it _specifically_ put inject-initialized properties first, but you would be able to configure it to put properties before getters and methods.

Thanks for the suggestion! But the bigger reason I think this belongs in `angular-eslint` specifically is that this footgun _didn't exist before `inject()`_.

With _constructor-parameter DI_, parameters are assigned _before any field initializer runs_, so injected dependencies were always available by the time the first field executed. You literally could not hit this.
The moment dependencies move into field initializers via `inject()`, the order of the class body starts to matter, and you inherit the full class-fields ordering footgun with it.

And that's _exactly the pattern [prefer-inject](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-inject.md) pushes people toward_. It's in recommended, so any project following this plugin's defaults is being nudged, by this plugin, into the shape of code where this bug becomes possible.

If `angular-eslint` ships the rule that _moves people onto `inject()` by default_, it feels consistent for it to also ship the rule that _prevents the main sharp edge that pattern introduces_. Otherwise recommended is, in effect, guiding users into a shape of code where a subtle runtime bug becomes reachable, without giving them the lint-level guardrail to catch it.
